### PR TITLE
tessel rename

### LINKED ISSUE #77: t2 rename
Current status: no such command

Spec: [T2 CLI spec](https://forums.tessel.io/t/t2-cli-spec-comments-requested/1743)
**tessel rename <name> [-r | --reset]**

Names the Tessel by changing the hostname of the device. Name restrictions found here: http://en.wikipedia.org/wiki/Hostname#Restrictions_on_valid_host_names
```sh
$ tessel rename
> Usage: tessel rename <name> [-r | --reset] [--help]
>  name:		the name to rename the currently selected tessel
>  -r | --reset:	reset the name of the device
$ 

$ tessel rename Jigglypuff
> Changed name of device Tessel-0123456789ab to Jigglypuff
$ 

$ tessel rename Snorlax
> Changed name of device Jigglypuff to Snorlax
$ 

$ tessel rename --reset
> Reset the name of device to Tessel-0123456789ab
$ 
```

* [ ] rename
* [ ] reset name to MAC address
* [ ] make sure hostname is valid based on hostname restrictions

### PR DESCRIPTION
**Pull request by johnnyman727**
Fixes #77

### CODE REVIEW

**Comment by johnnyman727** [comment]
This PR is nearly ready. I still need to figure out how to properly reset mdns with the new name.

**Comment by johnnyman727** [comment]
Ready for review. @Frijol?

File: bin/tessel-2.js [hunk 1]
```diff
@@ -200,4 +200,29 @@ parser.command('key')
     });
   })
 
+parser.command('rename')
+  .option('newName', {
+    metavar : 'newName',
```
Comment (Frijol) [inline bin/tessel-2.js]: other metavars have all been all caps

File: bin/tessel-2.js [hunk 2]
```diff
@@ -200,4 +200,29 @@ parser.command('key')
...
+    .then(function() {
+      process.exit(0);
+    })
+    .catch(function (err) {
+      if(err instanceof Error){
+        throw err;
+      }
+      tessel.logs.err(err);
+      process.exit(1);
+    });
```
Comment (Frijol) [inline bin/tessel-2.js]: add a .help for the rename command

File: lib/tessel/name.js [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+  var self = this;
+
+  // Regex to test whether a string is a valid Linux hostname
+  // Picked up from http://stackoverflow.com/questions/106179/regular-expression-to-match-dns-hostname-or-ip-address
+  var validHostnameRegex = /^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])$/;
+
+  // If the provided name isn't a string or a valid hostname
+  if (typeof name != 'string' || !validHostnameRegex.test(name)) {
+    // throw an error
+    return callback && callback("Invalid name: " + name + ". It must be a valid hostname string. See http://en.wikipedia.org/wiki/Hostname#Restrictions_on_valid_host_names.");
```
Comment (Frijol) [inline lib/tessel/name.js]: Weird to use "It" rather than a definite noun

**Comment by Frijol** [comment]
seems like it would be a good idea to have someone else take a look at this as well

File: lib/controller.js [hunk 4]
```diff
@@ -20,9 +20,18 @@ function listTessels(opts) {
...
+              return new Promise(function(resolve, reject) {
+                var t = new Tessel(usb_connection);
+                t.getName(function(err) {
+                  if (err) {
+                    return reject(err);
+                  }
+                  // Pass connections through for consistency
+                  tessel.logs.info('USB CONNECTION ', 'name: ', t.name);
+                   // Pass connections through for consistency
+                  resolve(usb_connection);
```
Comment (nplus11) [inline lib/controller.js]: `return resolve(usb_connection);`?

**Comment by johnnyman727** [comment]
@Frijol I addressed your comments. @nplus11 want to also give a review?

File: bin/tessel-2.js [hunk 5]
```diff
@@ -200,4 +200,29 @@ parser.command('key')
...
+  })
+  .help("Change the friendly identifier of a Tessel to something new.");
```
Comment (Frijol) [inline bin/tessel-2.js]: Really vague. Also probably useful to mention that this is in fact the hostname

**Comment by nplus11** [comment]
lgtm

File: bin/tessel-2.js (hunk 5 above)
Comment (Frijol) [inline bin/tessel-2.js]: alternately: "Change the name of a Tessel"

**Pull request merged**